### PR TITLE
ec2_snapshot: Add check_mode support to ec2_snapshot

### PR DESCRIPTION
**Pull request by mandar242**
##### SUMMARY

Add check_mode support to ec2_snapshot.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ec2_snapshot

### CODE REVIEW

File: tests/integration/targets/ec2_snapshot/tasks/main.yml [hunk 1]
```diff
@@ -60,23 +56,27 @@
               volume_size: 8
               delete_on_termination: true
       register: instance
+      retries: 5
+      delay: 60
+      until: "instance.instances[0].block_device_mappings | length > 0 and instance.instances[0].block_device_mappings[-1].ebs.status =='attached'"
```
Comment (tremble) [inline tests/integration/targets/ec2_snapshot/tasks/main.yml:61]: it should be possible to do this with `state: running` rather than retries.

File: tests/integration/targets/ec2_snapshot/tasks/main.yml [hunk 2]
```diff
@@ -222,11 +220,8 @@
           "tag:Name": '{{ resource_prefix }}'
       register: info_result
 
-    - assert:
-        that:
-          - info_result.snapshots| length == 3
-
-    - ec2_snapshot:
```
Comment (tremble) [inline tests/integration/targets/ec2_snapshot/tasks/main.yml:229]: Why remove this assertion?  Specifically with the addition of check_mode support it's important to know that we only created the expected number of snapshots.

File: plugins/modules/ec2_snapshot.py [hunk 3]
```diff
@@ -321,6 +318,8 @@ def create_snapshot(module, ec2, description=None, wait=None,
 
 
 def delete_snapshot(module, ec2, snapshot_id):
+    if module.check_mode:
+        module.exit_json(changed=True, msg='Would have deleted snapshot if not in check mode')
```
Comment (tremble) [inline plugins/modules/ec2_snapshot.py:322]: The existing code uses the `InvalidSnapshot.NotFound` exception to tell if the snapshot existed or not.  This will break idempotency.

Since the integration tests didn't catch this it would be good to add a test that running deletion twice results in `changed` followed by  `not changed`

File: plugins/modules/ec2_snapshot.py [hunk 4]
```diff
@@ -71,39 +71,33 @@
...
 - amazon.aws.ec2
-
 '''
 
 EXAMPLES = '''
 # Simple snapshot of volume using volume_id
 - amazon.aws.ec2_snapshot:
     volume_id: vol-abcdef12
     description: snapshot of /data from DB123 taken 2013/11/28 12:18:32
-
```
Comment (tremble) [inline plugins/modules/ec2_snapshot.py:87]: Do we have any guidance around the whitespace in here? Personally I find it easier to scan-read with the whitespace between examples

File: plugins/modules/ec2_snapshot.py [hunk 5]
```diff
@@ -364,6 +363,7 @@ def create_snapshot_ansible_module():
         required_if=required_if,
         required_one_of=required_one_of,
         required_together=required_together,
+        supports_check_mode=True
```
Comment (tremble) [inline plugins/modules/ec2_snapshot.py:366]:
```suggestion
        supports_check_mode=True,
```

If you add the extra `,` on the end of the line here, then when someone adds another entry they don't need to change an unrelated line.

File: changelogs/fragments/512-ec2_snapshot_add_check_mode_support.yml.yml [hunk 6]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - ec2_snapshot - add check_mode support
+    (https://github.com/ansible-collections/amazon.aws/pull/512)
```
Comment (tremble) [inline changelogs/fragments/512-ec2_snapshot_add_check_mode_support.yml.yml:3]:
```suggestion
    (https://github.com/ansible-collections/amazon.aws/pull/512).
```

Very much just a niggle, but is a minor style issue.  https://docs.ansible.com/ansible/latest/community/development_process.html#changelog-fragment-entry-format

**Review by tremble — CHANGES_REQUESTED** [review]
Mostly just minor niggles.  Thanks for your work on this.

File: tests/integration/targets/ec2_snapshot/tasks/main.yml [hunk 7]
```diff
@@ -59,24 +55,27 @@
             ebs:
               volume_size: 8
               delete_on_termination: true
+        state: running
       register: instance
+      until: "instance.instances[0].block_device_mappings | length > 0 and instance.instances[0].block_device_mappings[-1].ebs.status =='attached'"
```
Comment (tremble) [inline tests/integration/targets/ec2_snapshot/tasks/main.yml:60]:
```suggestion
        state: running
        wait: true
      register: instance
```

**Review by tremble — CHANGES_REQUESTED** [review]

**Review by jillr — APPROVED** [review]
LGTM, I'll leave gate off for now to give @tremble a chance to update his review for the latest changes he requested. Thanks @mandar242!

File: tests/integration/targets/ec2_snapshot/tasks/main.yml [hunk 8]
```diff
@@ -295,12 +292,36 @@
...
+      ec2_snapshot:
+        state: absent
+        snapshot_id: '{{ tagged_snapshot_id }}'
+      register: delete_result_second_check_mode
+      check_mode: true
+
+    - assert:
+        that:
+          - delete_result_second_check_mode is not changed
+
```
Comment (tremble) [inline tests/integration/targets/ec2_snapshot/tasks/main.yml:324]: Minor niggle: please add a check for deleting an already removed snapshot not in checkmode.  If it errors just comment it out and open an issue.

**Review by tremble — CHANGES_REQUESTED** [review]
Thanks,

I'd still like to see a non-check-mode deletion idempotency test (ie deleting a deleted snapshot doesn't error), but otherwise looks good.

**Review by tremble — APPROVED** [review]
LGTM.  Let's get this gated!

**Comment by tremble** [comment]
check had passed, I accidentally closed the PR.

**Comment by tremble** [comment]
Check buildset: https://dashboard.zuul.ansible.com/t/ansible/buildset/a2cfdceedb3d442a889a32087b755dd5
